src/Compile/CompileEnv.ts

```diff
@@ -0,0 +1,169 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import assert from 'assert';
+
+import {Compiler} from '../Backend/Compiler';
+import {Toolchain} from '../Backend/Toolchain';
+import {BuilderJob} from '../Project/BuilderJob';
+import {Job} from '../Project/Job';
+import {JobConfig} from '../Project/JobConfig';
+import {JobInstall} from '../Project/JobInstall';
+import {JobInstalled} from '../Project/JobInstalled';
+import {JobUninstall} from '../Project/JobUninstall';
+import {WorkFlow} from '../Project/WorkFlow';
+import {Balloon} from '../Utils/Balloon';
+import * as helpers from '../Utils/Helpers';
+import {Logger} from '../Utils/Logger';
+
+class Env implements BuilderJob {
+  workFlow: WorkFlow;  // our build WorkFlow
+  currentWorkspace: string = '';
+  isPrepared: boolean = false;
+
+  constructor(l: Logger) {
+    this.workFlow = new WorkFlow(l);
+  }
+
+  public init() {
+    this.workFlow.clearJobs();
+  }
+
+  public addJob(job: Job): void {
+    assert(this.isPrepared === false);
+    this.workFlow.addJob(job);
+  }
+
+  public clearJobs(): void {
+    this.workFlow.clearJobs();
+    this.isPrepared = false;
+  }
+
+  public finishAdd(): void {
+    console.log('Done building WorkFlow on Env');
+    console.log(this.workFlow.jobs);
+    this.isPrepared = true;
+  }
+
+  public build() {
+    if (this.isPrepared !== true) {
+      throw Error('Env is not yet prepared');
+    }
+
+    try {
+      this.currentWorkspace = helpers.obtainWorkspaceRoot();
+    } catch (e: unknown) {
+      let errmsg = 'Failed to obtain workspace root';
+      if (e instanceof Error) {
+        errmsg = e.message;
+      }
+      // TODO add more type for e if changed in obtainWorkspaceRoot
+      Balloon.error(errmsg);
+      return;
+    }
+
+    this.workFlow.start(this.currentWorkspace);
+  }
+}
+
+// A CompileEnv has a Toolchain
+class CompileEnv extends Env {
+  installed?: Toolchain;
+  compiler: Compiler;
+
+  constructor(l: Logger, compiler: Compiler) {
+    super(l);
+    this.installed = undefined;
+    this.compiler = compiler;
+    this.init();
+  }
+
+  confirmInstalled() {
+    this.clearJobs();
+
+    const toolchains = this.compiler.toolchains();
+    for (let index = 0; index < toolchains.length; index++) {
+      const toolchain = toolchains[index];
+      let cmd = toolchain.installed();
+      let job = new JobInstalled(cmd);
+      job.successCallback = () => {
+        this.installed = toolchain;
+      };
+      this.addJob(job);
+    }
+
+    this.finishAdd();
```

Comment:
If `toolchains()` returns long list of toolchains, this will take some time. (one command per one toolchain version)
I wonder if there any shell command which returns installed version of toolchain. If so, we can compare the version for each toolchain returned from `toolchains()`.

Reply:
This code searches for all the packages on the list and the installed package that is searched last is registered as an `installed` package. The reason for this implementation is that the debian package is installed only one version in one environment. We previously shared that is was necessary to change this, but we decided to keep this implementation as it is until demo.